url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/com/wroblicky/andrew/joust/game/chesspiece/Pawn.java

```diff
@@ -0,0 +1,184 @@
+package com.wroblicky.andrew.joust.game.chesspiece;
+
+import static com.wroblicky.andrew.joust.game.chesspiece.ChessPiece.Occupier.ENEMY;
+import static com.wroblicky.andrew.joust.game.chesspiece.ChessPiece.Occupier.UNOCCUPIED;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import com.wroblicky.andrew.joust.game.board.ChessBoard;
+import com.wroblicky.andrew.joust.game.board.Location;
+import com.wroblicky.andrew.joust.game.subset.qualifiable.ChessPieceAllegianceType;
+import com.wroblicky.andrew.joust.game.subset.qualifiable.ChessPieceType;
+
+/**
+ * Represents the pawn chess piece
+ *
+ * @author Andrew Wroblicky
+ *
+ */
+public final class Pawn extends ChessPiece {
+
+	private boolean hasMovedAtAll;
+	private boolean movedTwice;
+
+	public Pawn() {
+		this.allegiance = Allegiance.BLACK;
+		this.hasMovedAtAll = false;
+		this.movedTwice = false;
+		this.alive = true;
+		this.chessID = "p1";
+		this.chessBoard = new ChessBoard();
+	}
+
+	public Pawn(Allegiance allegiance, int chessID, ChessBoard chessBoard) {
+		this.allegiance = allegiance;
+		this.hasMovedAtAll = false;
+		this.movedTwice = false;
+		this.alive = true;
+		this.chessID = getChessPieceAllegianceType()
+				.getChessPieceAllegianceType() + chessID;
+		this.chessBoard = chessBoard;
+	}
+
+	public boolean isMovedAtAll() {
+		return hasMovedAtAll;
+	}
+
+	public boolean isMovedTwice() {
+		return movedTwice;
+	}
+
+	@Override
+	public ChessPieceType getChessPieceType() {
+		return ChessPieceType.PAWN;
+	}
+
+	@Override
+	public ChessPieceAllegianceType getChessPieceAllegianceType() {
+		if (isBlack()) {
+			return ChessPieceAllegianceType.BLACK_PAWN;
+		}
+		return ChessPieceAllegianceType.WHITE_PAWN;
+	}
+
+	public void setMovedAtAll(boolean hasMovedAtAll) {
+		this.hasMovedAtAll = hasMovedAtAll;
+	}
+
+	public void setMovedTwice(boolean movedTwice) {
+		this.movedTwice = movedTwice;
+	}
+
+	// TODO
+	// add support for pawn promotion
+	@Override
+	public List<Location> getPossibleMoves() {
+		List<Location> possibles = new ArrayList<Location>();
+		Location moveUpLeft = getWestAttackNeighbor();
+		Location moveUpRight = getEastAttackNeighbor();
+		Location left = chessBoard.getWestNeighbor(getLocation());
+		Location right = chessBoard.getEastNeighbor(getLocation());
+
+		// forward moves
+		possibles = getForwardMoves(possibles);
+
+		// attack moves
+		if (moveUpLeft != null && checkAvailability(moveUpLeft) == ENEMY) {
+			possibles.add(moveUpLeft);
+		}
+		if (moveUpRight != null && checkAvailability(moveUpRight) == ENEMY) {
+			possibles.add(moveUpRight);
+		}
+
+		// en passant
+		if (left != null && checkAvailability(left) == ENEMY
+				&& checkAvailability(moveUpLeft) == UNOCCUPIED) {
+			ChessPiece enemy = chessBoard.getChessPieceByLocation(left);
+			if (enemy instanceof Pawn) {
+				Pawn enemyPawn = (Pawn) enemy;
+				if (enemyPawn.isMovedTwice()
+						&& enemyPawn.getRelativeRank().equals("4")) {
+					possibles.add(moveUpLeft);
+				}
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 95:105, 107:117.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.